Async result type for results inside of futures

janstenpickle opened the pull request:
Raising this more so we can have a bit of a discussion on this; whether it is sensible, and if so how the implementation should look.

This is very loosely modelled on scalaz's `Task`, however doesn't provide quite so much functionality. Really it's just a wrapper around `Future[Result[A, B]`. Most of the functionality delegates to the underlying `Result` which is inside of the future. It does provide a `flatMap` instance which allows it to be used in a for comprehension. 

Wanted to discuss if.
- The functionality makes sense
- If the functionality make sense, then do the names make sense?
- Does any more functionality need adding, like `onComplete`, `onSuccess` or `onFail`? Or whether we should just say that people should access that via the future in the `underlying` parameter (similar to the scalaz `Task` `get`?

Commit:
Add async result type

albertpastrana commented:
Hi @janstenpickle, thanks again for the contribution, it's great!

Regarding your comments/questions:

> - The functionality makes sense

Yes, it does!

> - If the functionality make sense, then do the names make sense?

I like a lot `AsyncResult`, I think it's better than `FutureResult` or `Task` or anything else I can think of. The only thing I can think of is that people is used to the `Future` nomenclature and may be a bit difficult to find.
Regarding the package, I'm not sure but we may want to put it into `uscala.concurrent` or `uscala.async` instead of adding a new level?

> -Does any more functionality need adding, like onComplete, onSuccess or onFail? Or whether we should just say that people should access that via the future in the underlying parameter (similar to the scalaz Task get?

I think for now is more than fine and using the `underlying` future should be enough for most use cases. And if somebody really wants to have these methods in here, they can raise a pull request :-)

I have some small comments on the implementation that I'll leave inline.

albertpastrana commented on result-async/src/main/scala/uscala/result/async/AsyncResult.scala:
```diff
@@ -0,0 +1,89 @@
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.Try
+
+final case class AsyncResult[+A, +B](underlying: Future[Result[A, B]]) {
```
Not sure about it being a case class, given the implementation I would have probably chosen a "regular" class.

albertpastrana commented on result-async/src/main/scala/uscala/result/async/AsyncResult.scala:
```diff
@@ -0,0 +1,89 @@
+      )
+    )
+
+
```
extra space in here :)

albertpastrana commented on result-async/src/main/scala/uscala/result/async/AsyncResult.scala:
```diff
@@ -0,0 +1,89 @@
+
+  def fold[C](fa: A => C, fb: B => C)(implicit ec: ExecutionContext): Future[C] = underlying.map(_.fold(fa, fb))
+
+  def attemptRun[AA >: A](implicit f: Throwable => AA, ec: ExecutionContext): Result[AA, B] =
```
Not sure if I like the function that maps the `Throwable` to be passed implicitly. I see it more like the `Result.fromOption(o: Option[B], ifEmpty: => A)` function?

albertpastrana commented on result-async/src/main/scala/uscala/result/async/AsyncResult.scala:
```diff
@@ -0,0 +1,89 @@
+  def attemptRun[AA >: A](implicit f: Throwable => AA, ec: ExecutionContext): Result[AA, B] =
+    Result.fromTry(
+      Try(Await.result(underlying, Duration.Inf))
+    ).leftMap(f).flatMap(identity)
```
You could use `Result.attempt` in here:

``` scala
Result.attempt(Await.result(underlying, Duration.Inf)).mapFail(f).flatMap(identity)
```

Also, I'm wandering if we should provide two different methods:

The first one does not flatMap the results so you can see what has been wrong, the `Future` or the computation:

``` scala
def attemptRun(implicit ec: ExecutionContext): Result[Throwable, Result[A, B]] =
  Result.attempt(Await.result(underlying, Duration.Inf))
```

The second does allow you to transform the `Future` failure into your own error type:

``` scala
def attemptRun[AA >: A](f: Throwable => AA)(implicit ec: ExecutionContext): Result[AA, B] =
  Result.attempt(Await.result(underlying, Duration.Inf)).mapFail(f).flatMap(identity)
```

janstenpickle replied:
Yep, I think that makes sense having both, I'll make the change

albertpastrana commented on result-async/src/main/scala/uscala/result/async/AsyncResult.scala:
```diff
@@ -0,0 +1,89 @@
+
+  def fromFuture[A, B](f: Future[B])(implicit ex: ExecutionContext): AsyncResult[A, B] = AsyncResult(f.map(Result.ok))
+
+  def fromResult[A, B](r: Result[A, B])(implicit ex: ExecutionContext): AsyncResult[A, B] = AsyncResult(Future(r))
```
This should use the `Future.successful` method. That way you don't need the `ExecutionContext` as it's just a completed `Promise`.

albertpastrana commented on result-async/src/main/scala/uscala/result/async/AsyncResult.scala:
```diff
@@ -0,0 +1,89 @@
+
+  def fromResult[A, B](r: Result[A, B])(implicit ex: ExecutionContext): AsyncResult[A, B] = AsyncResult(Future(r))
+
+  def now[A, B](b: B)(implicit ex: ExecutionContext): AsyncResult[A, B] = AsyncResult(Future(Result.ok(b)))
```
I see this method more like:

``` scala
def now[A, B](r: Result[A, B]): AsyncResult[A, B] = AsyncResult(Future.successful(r))
```

janstenpickle replied:
That functionality is already in `fromResult`. I was trying to mimic the scalaz `Task` functionality, however it might be better to just remove it entirely to avoid any confusion with behaviour.

albertpastrana commented on result-async/src/main/scala/uscala/result/async/AsyncResult.scala:
```diff
@@ -0,0 +1,89 @@
+
+  def ok[A, B](b: B)(implicit ex: ExecutionContext): AsyncResult[A, B] = now(b)
+
+  def fail[A, B](a: A)(implicit ex: ExecutionContext): AsyncResult[A, B] = AsyncResult(Future(Result.fail(a)))
```
Given the above definition of `now`, the `ok` and `fail` methods could be rewritten as:

``` scala
def ok[A, B](b: B): AsyncResult[A, B] = now(Result.ok(b))

def fail[A, B](a: A): AsyncResult[A, B] = now(Result.fail(b))
```

albertpastrana commented on result-async/src/main/scala/uscala/result/async/AsyncResult.scala:
```diff
@@ -0,0 +1,89 @@
+  def leftMap[C](f: A => C)(implicit ec: ExecutionContext): AsyncResult[C, B] =
+    AsyncResult(
+      underlying.map(
+        _.leftMap(f)
```
I prefer using `mapFail` as it's more intuitive imo ;-)

albertpastrana commented on result-async/src/main/scala/uscala/result/async/AsyncResult.scala:
```diff
@@ -0,0 +1,89 @@
+    ).leftMap(f).flatMap(identity)
+}
+
+object AsyncResult {
```
Maybe we could add this method?

``` scala
def attempt[B](f: => B)(implicit ex: ExecutionContext): AsyncResult[Throwable, B] = AsyncResult(Future(Result.attempt(f)))
```

albertpastrana commented on result-async/src/main/scala/uscala/result/async/AsyncResult.scala:
```diff
@@ -0,0 +1,89 @@
+import scala.util.Try
+
+final case class AsyncResult[+A, +B](underlying: Future[Result[A, B]]) {
+  def flatMap[AA >: A, D](f: B => AsyncResult[AA, D])(implicit ec: ExecutionContext): AsyncResult[AA, D] =
```
Need to think properly about this method :)

albertpastrana commented on result-async/src/main/scala/uscala/result/async/AsyncResult.scala:
```diff
@@ -0,0 +1,89 @@
+    AsyncResult(
+      map(f).underlying.flatMap(
+        _.fold(
+          err => Future(Result.fail(err)),
```
You could use `Future.successful` or even `AsynResult.fail(err).underlying`

Commit:
PR changes and test for AsyncResult

- used `Future.success` where neccessary
- removed confusing `.now` syntax
- added `attempt` syntax
- added new `attemptRun` and `attemptRunFor` overloaded methods
- reworked `flatMap` implementation
- added tests

janstenpickle commented:
@albertpastrana I've made changes as per your suggestions and added tests.

With regard to naming, I see your point - I was going to say `AsyncResult` is no more confusing than `Task` in scalaz, but I understand that part of the point of this microlib is to be less confusing than scalaz! So I'm not sure what to call it, `FutureResult` maybe?
